bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -0,0 +1,321 @@
+import jsky.app.ot.OT
+import jsky.app.ot.gemini.altair.Altair_WFS_Feature
+import jsky.app.ot.gemini.inst.OIWFS_Feature
+import jsky.app.ot.gemini.tpe.TpePWFSFeature
+import jsky.app.ot.tpe.{GemsGuideStarWorker, GuideStarSupport, TpeContext, TpeManager}
+import scala.collection.JavaConverters._
+import scala.collection.immutable.HashMap
+import scala.swing.Swing
+import scala.util.{Try, Failure, Success}
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+object BagsManager {
+  private val LOG: Logger = Logger.getLogger(getClass.getName)
+
+  private case class BagsTask(observation: ISPObservation, curCtx: Option[ObsContext], nextCtx: Option[ObsContext]) {
+    def obsKey: SPNodeKey =
+      observation.getNodeKey
+
+    // The state to move to when processing the task.
+    def nextTask: Option[BagsTask] =
+      nextCtx.map(_ => BagsTask(observation, nextCtx, None))
+
+    def isDone: Boolean =
+      nextCtx.isEmpty
+
+    def newContext(ctxOpt: Option[ObsContext]): BagsTask =
+      BagsTask(observation, curCtx, ctxOpt)
+
+    def performLookup(): Unit = {
+      LOG.info(s"Performing BAGS lookup for observation=${observation.getObservationID}")
+
+
+      def resultString[A](r: Option[A]): String =
+        s"Results=${r.fold("No")(_ => "Yes")}."
+
+      // We must handle GeMS separately from other strategies, and delegate the work to the GemsGuideStarWorker.
+      def gemsLookup(obsCtx: ObsContext): Unit = {
+        LOG.info(s"BAGS GeMS lookup for observation=${observation.getObservationID} starting...")
+        val ggsTry = Try { GemsGuideStarWorker.findGuideStars(obsCtx) }
+        LOG.info(s"BAGS GeMS lookup for observation=${observation.getObservationID} ${ggsTry.map(ggs => s"successful. " +
+          resultString(ggs.asScalaOpt)).getOrElse("failed.")}")
+        if (ggsTry.isFailure)
+          LOG.warning(s"EXCEPTION=${ggsTry.failed.get}")
+        ggsTry.foreach { ggs =>
+          Swing.onEDT {
+            muteObservation(observation)
+            GemsGuideStarWorker.applyResults(TpeContext(observation), ggs, true)
+            unmuteObservation(observation)
+          }
+        }
+        taskComplete(this, success = ggsTry.isSuccess)
+      }
+
+      // The mechanism to perform lookups for all other guide probes.
+      def otherLookup(obsCtx: ObsContext, strategy: AgsStrategy): Unit = {
+        def showTpeFeatures(selOpt: Option[AgsStrategy.Selection]): Unit =
+          selOpt.foreach { sel =>
+            Option(TpeManager.get()).filter(_.isVisible).foreach { tpe =>
+              sel.assignments.foreach { ass =>
+                val clazz = ass.guideProbe.getType match {
+                  case GuideProbe.Type.AOWFS => classOf[Altair_WFS_Feature]
+                  case GuideProbe.Type.OIWFS => classOf[OIWFS_Feature]
+                  case GuideProbe.Type.PWFS => classOf[TpePWFSFeature]
+                }
+                Option(tpe.getFeature(clazz)).foreach {
```

Comment:
I wish we could get away from the reflection based properties of the TPE, probably it doesn't belong to this PR though